portal/src/graphql/portal/SingleSignOnConfigurationScreen.tsx

```diff
@@ -0,0 +1,375 @@
+import React, { useCallback, useMemo, useRef, useState } from "react";
+import { useParams } from "react-router-dom";
+import { produce } from "immer";
+import { Link, Text } from "@fluentui/react";
+import { FormattedMessage } from "@oursky/react-messageformat";
+
+import SingleSignOnConfigurationWidget from "./SingleSignOnConfigurationWidget";
+import ShowLoading from "../../ShowLoading";
+import ShowError from "../../ShowError";
+import ButtonWithLoading from "../../ButtonWithLoading";
+import { useAppAndSecretConfigQuery } from "./query/appAndSecretConfigQuery";
+import { useUpdateAppAndSecretConfigMutation } from "./mutations/updateAppAndSecretMutation";
+import { clearEmptyObject, ensureNonEmptyString } from "../../util/misc";
+import { parseError } from "../../util/error";
+import { Violation } from "../../util/validation";
+import {
+  OAuthClientCredentialItem,
+  OAuthSSOProviderConfig,
+  OAuthSSOProviderType,
+  oauthSSOProviderTypes,
+  PortalAPIApp,
+  PortalAPIAppConfig,
+  PortalAPISecretConfig,
+} from "../../types";
+
+import styles from "./SingleSignOnConfigurationScreen.module.scss";
+
+interface SingleSignOnConfigurationProps {
+  rawAppConfig: PortalAPIAppConfig | null;
+  effectiveAppConfig: PortalAPIAppConfig | null;
+  secretConfig: PortalAPISecretConfig | null;
+  updatingAppConfig: boolean;
+  updateAppConfig: (
+    appConfig: PortalAPIAppConfig,
+    secretConfig: PortalAPISecretConfig
+  ) => Promise<PortalAPIApp | null>;
+  updateAppConfigError: unknown;
+}
+
+export interface OAuthSSOProviderConfigState extends OAuthSSOProviderConfig {
+  enabled: boolean;
+  clientSecret?: string;
+}
+
+export type SingleSignOnScreenState = {
+  [key in OAuthSSOProviderType]: OAuthSSOProviderConfigState;
+};
+
+function getProviderConfigStateFromProviders(
+  providerType: OAuthSSOProviderType,
+  providerByType: Partial<Record<OAuthSSOProviderType, OAuthSSOProviderConfig>>,
+  clientSecretByAlias: Partial<Record<string, string>>
+): OAuthSSOProviderConfigState {
+  const providerConfig = providerByType[providerType] ?? {
+    type: providerType,
+  };
+  // FIXME: make alias required
+  const alias = providerConfig.alias ?? providerType;
+  return {
+    ...providerConfig,
+    enabled: !!providerByType[providerType],
+    clientSecret: clientSecretByAlias[alias],
+  };
+}
+
+function extractOAuthSecret(secretConfig: PortalAPISecretConfig) {
+  for (const secret of secretConfig.secrets) {
+    if (secret.key === "sso.oauth.client") {
+      return secret;
+    }
+  }
+  return null;
+}
+
+function constructScreenState(
+  providers: OAuthSSOProviderConfig[],
+  oauthCredentials: OAuthClientCredentialItem[]
+) {
+  const providerByType = providers.reduce(
```

Comment:
tsserver complains that `Object.fromEntries` does not exist, shall we use newer target, say `es2019` ? @kiootic

Reply:
According to [ES compatibility table](https://kangax.github.io/compat-table/es2016plus/), we can use up to ES2019 for most of our supported browsers.

I suppose we can remove Android/iOS browser from `.browserlistrc`, since we don't really support mobile platform for portal. For Desktop Safari, we may want to bump it to >= 12. Otherwise, we may just use core-js for polyfills.

@louischan-oursky any suggestion?

Reply:
In `index.tsx`, we have bundled core-js. So we can use all stage-4 features of ECMAScript.

Let's update tsconfig.json to make it `"lib": ["DOM", "ES5", "ES2015"]`

Reply:
@louischan-oursky how about use `ES2019` instead? Most major browsers supports that, and we may want to bump supported Safari version to latest 2 version to match [Fluent UI's browser support](https://github.com/microsoft/fluentui/wiki/Browser-Support).